### PR TITLE
Reorganize operations

### PR DESCRIPTION
**Pull request by breuleux**
This is the beginning of an attempt to reorganize all operations in Myia. This sets up a new system that cohabits with the old system, for now.

* All operations, whether they be macros, composites or primitives, would be moved to `myia.operations`
* Each primitive would have its own file, e.g. `scalar_pow` would live in `myia.operations.prim_scalar_pow` (`scalar_pow` is the only primitive to move to the new system at the moment).
* `myia.operations.xyz` is an instance of `Operation`, which optionally defines a Python implementation and a default mapping (which can be a Primitive, a Macro, etc.) The default mapping can be overridden in the object_map.
* A `Primitive` now defines an optional Python implementation (for the debug VM), a default inferrer, and a default bprop, which can all be overridden.

Under the new system, the following files would disappear:

* `myia.prim.ops`: now `myia.operations.primitives`
* `myia.prim.py_implementations`: import from `myia.operations` instead.
* `myia.prim.grad_implementations`
* `myia.abstract.prim`
* `myia.composite`, now in individual modules under `myia.operations`
* `myia.macros`, now in individual modules under `myia.operations`

### CODE REVIEW

File: myia/compile/backends/nnvm.py [hunk 1]
```diff
@@ -411,3 +411,22 @@ def from_scalar(self, s, t):
 def NNVMBackendR(target, device_id):
     """NNVM Proxy."""
     return HandleBackend(NNVMBackend(target, device_id))
+
+
+__all__ = [
+    'NNVMBackend',
+    'NNVMBackendR',
```
Comment (abergeron) [inline myia/compile/backends/nnvm.py]: This should only export NNVMBackend and NNVMBackendR

File: myia/compile/backends/pytorch.py [hunk 2]
```diff
@@ -288,3 +288,21 @@ def from_scalar(self, s, t):
 def PyTorchBackendR(device):
     """Pytorch proxy."""
     return HandleBackend(PyTorchBackend(device))
+
+
+__all__ = [
+    'PyTorchBackend',
+    'PyTorchBackendR',
```
Comment (abergeron) [inline myia/compile/backends/pytorch.py]: Same here

File: myia/compile/backends/relay.py [hunk 3]
```diff
@@ -400,3 +400,21 @@ def from_scalar(self, s, t):
 def RelayBackendR(target, device_id):
     """Relay proxy."""
     return HandleBackend(RelayBackend(target, device_id))
+
+
+__all__ = [
+    'CompileGraph',
+    'MAP',
+    'RelayBackend',
+    'RelayBackendR',
```
Comment (abergeron) [inline myia/compile/backends/relay.py]: And here.

File: myia/compile/vm.py [hunk 4]
```diff
@@ -361,3 +361,9 @@ def inst_external(self, fn, args):
         outs = fn(*(self._ref(a) for a in args))
         for o in outs:
             self._push(o)
+
+
+__all__ = [
+    'FinalVM',
+    'struct_partial',
```
Comment (abergeron) [inline myia/compile/vm.py]: No struct_partial export
```suggestion
```

File: myia/frontends/pytorch.py [hunk 5]
```diff
@@ -183,3 +183,9 @@ def _to_abstract(self, v: torch.nn.Parameter, **kwargs):
         }),
         {SHAPE: tuple(v.shape), TYPE: PyTorchTensor},
     )
+
+
+__all__ = [
+    'mod_sub',
+    'pytorch_dtype_to_type',
+]
```
Comment (abergeron) [inline myia/frontends/pytorch.py]: I would argue that this shouldn't have an `__all__`

File: myia/grad.py [hunk 6]
```diff
@@ -623,3 +623,18 @@ def bprop_to_grad_transform(prim, **flags):
...
+    'BPropAppRemapper',
+    'BPropRemapper',
+    'FPropAppRemapper',
+    'FPropRemapper',
+    'GradRemapper',
+    'J',
+    'SensRemapper',
+    'SlaveRemapper',
+    'bprop_to_grad_transform',
+    'wrap_grad_transform',
```
Comment (abergeron) [inline myia/grad.py]: This might be a bit too much exporting.

File: myia/graph_utils.py [hunk 7]
```diff
@@ -102,3 +102,12 @@ def toposort(root: T,
         done.add(node)
         yield node
         todo.pop()
+
+
+__consolidate__ = True
+__all__ = [
+    'T',
```
Comment (abergeron) [inline myia/graph_utils.py]: Not this.
```suggestion
```

File: myia/info.py [hunk 8]
```diff
@@ -206,3 +206,14 @@ def __exit__(self, type, value, tb):
...
+
+
+__consolidate__ = True
+__all__ = [
+    'About',
+    'DebugInfo',
+    'DebugInherit',
+    'NamedDebugInfo',
+    'StackVar',
+    'current_info',
```
Comment (abergeron) [inline myia/info.py]: Maybe current_info and StackVar are somewhat private.

File: myia/ir/manager.py [hunk 9]
```diff
@@ -897,3 +897,35 @@ def __enter__(self):
...
+    'NestingStatisticGraphWise',
+    'NestingStatisticWholesale',
+    'NodesStatistic',
+    'ParentProxy',
+    'ParentStatistic',
+    'PerGraphStatistic',
+    'RecursiveStatistic',
+    'ScopeStatistic',
+    'UsesStatistic',
+    'manage',
```
Comment (abergeron) [inline myia/ir/manager.py]: I think we only want manage and ManagerError here.

File: myia/opt/cse.py [hunk 10]
```diff
@@ -99,3 +99,10 @@ def __call__(self, root):
         """Apply CSE on root."""
         chg = cse(root, self.resources.manager)
         return chg and self.report_changes
+
+
+__all__ = [
+    'CSE',
+    'cse',
+    'group_nodes',
```
Comment (abergeron) [inline myia/opt/cse.py]: Perhaps not this.
```suggestion
```

File: myia/opt/lib.py [hunk 11]
```diff
@@ -1048,3 +1048,17 @@ def __call__(self, root):
             node.abstract = newtype
 
         return len(nodes) > 0
+
+
+__all__ = [
+    'JElim',
+    'M',
```
Comment (abergeron) [inline myia/opt/lib.py]: Maybe not this
```suggestion
```

File: myia/utils/env.py [hunk 12]
```diff
@@ -106,3 +106,13 @@ def __len__(self):
...
 newenv = EnvInstance()
+
+
+__consolidate__ = True
+__all__ = [
+    'EnvInstance',
+    'SymbolicKeyInstance',
+    'newenv',
+    'require_same',
+    'smap',
```
Comment (abergeron) [inline myia/utils/env.py]: Not this?

File: myia/utils/intern.py [hunk 13]
```diff
@@ -307,3 +307,26 @@ def intern(inst):
...
+    'PossiblyRecursive',
+    'RecursionException',
+    'Wrapper',
+    'deep_eqkey',
+    'eq',
+    'eqkey',
+    'eqrec',
+    'hash',
+    'hashrec',
+    'intern',
```
Comment (abergeron) [inline myia/utils/intern.py]: I'm pretty sure some of those should not be exported.

File: myia/utils/misc.py [hunk 14]
```diff
@@ -435,3 +435,31 @@ def assert_scalar(*args):
...
+    'TaggedValue',
+    'UNKNOWN',
+    'assert_scalar',
+    'core',
+    'dataclass_fields',
+    'is_dataclass_type',
+    'keyword_decorator',
+    'list_str',
+    'repr_',
+    'resolve_from_path',
```
Comment (abergeron) [inline myia/utils/misc.py]: A bunch of those should not be here.

File: myia/utils/trace.py [hunk 15]
```diff
@@ -335,3 +335,18 @@ def new_fn(**kwargs):
...
+__consolidate__ = True
+__all__ = [
+    'DoTrace',
+    'ProfileResults',
+    'Profiler',
+    'TraceExplorer',
+    'TraceListener',
+    'Tracer',
+    'TracerContextManager',
+    'glob_to_regex',
```
Comment (abergeron) [inline myia/utils/trace.py]: Maybe not this.  And some of the above ones too.

File: myia/utils/unify.py [hunk 16]
```diff
@@ -613,3 +613,25 @@ def reify(self, v, equiv: EquivT) -> Any:
...
+
+
+__consolidate__ = True
+__all__ = [
+    'FilterVar',
+    'PredicateSet',
+    'RestrictedVar',
+    'SVar',
+    'Seq',
+    'T',
```
Comment (abergeron) [inline myia/utils/unify.py]: Not this.  And perhaps some others too.

File: myia/utils/variables.py [hunk 17]
```diff
@@ -50,3 +50,9 @@ def _is_c(n):
 Xs = SVar(Var())
 Ys = SVar(Var())
 Cs = SVar(constvar())
+
+
+__consolidate__ = True
+__all__ = [
+    'constvar',
```
Comment (abergeron) [inline myia/utils/variables.py]: We might not want this, but all the other ones instead.

File: myia/vm.py [hunk 18]
```diff
@@ -290,3 +290,11 @@ def _handle_node(self, node: ANFNode, frame: VMFrame):
 
         else:
             raise AssertionError("Unknown node type")  # pragma: no cover
+
+
+__all__ = [
+    'Closure',
+    'Partial',
+    'VM',
+    'VMFrame',
```
Comment (abergeron) [inline myia/vm.py]: Probably not VMFrame.

File: myia/xtype.py [hunk 19]
```diff
@@ -318,3 +318,41 @@ class External(typing.Generic[T]):
...
+    'Float',
+    'Int',
+    'NDArray',
+    'Nil',
+    'NotImplementedType',
+    'Number',
+    'Object',
+    'String',
+    'SymbolicKeyType',
+    'T',
```
Comment (abergeron) [inline myia/xtype.py]: Not T, TypeMeta, as_frozen and maybe not Type too.

File: myia/frontends/pytorch.py (hunk 5 above)
Comment (breuleux) [reply myia/frontends/pytorch.py]: I'd keep `pytorch_dtype_to_type` I think.

File: myia/info.py (hunk 8 above)
Comment (breuleux) [reply myia/info.py]: Could be useful to export `current_info`

File: myia/utils/misc.py (hunk 14 above)
Comment (breuleux) [reply myia/utils/misc.py]: T1 and T2 don't belong, but the rest seems fine.

File: myia/utils/variables.py (hunk 17 above)
Comment (breuleux) [reply myia/utils/variables.py]: opt/lib does use constvar.

File: myia/xtype.py (hunk 19 above)
Comment (breuleux) [reply myia/xtype.py]: `TypeMeta` is used elsewhere.

**Pull request merged**